storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -253,17 +253,33 @@ public void append(long largestOffset,
             // append the messages
             long appendedBytes = log.append(records);
             LOGGER.trace("Appended {} to {} at end offset {}", appendedBytes, log.file(), largestOffset);
-            // Update the in memory max timestamp and corresponding offset.
-            if (largestTimestampMs > maxTimestampSoFar()) {
-                maxTimestampAndOffsetSoFar = new TimestampOffset(largestTimestampMs, shallowOffsetOfMaxTimestamp);
-            }
-            // append an entry to the index (if needed)
-            if (bytesSinceLastIndexEntry > indexIntervalBytes) {
-                offsetIndex().append(largestOffset, physicalPosition);
-                timeIndex().maybeAppend(maxTimestampSoFar(), shallowOffsetOfMaxTimestampSoFar());
-                bytesSinceLastIndexEntry = 0;
+
+            long batchMaxTimestamp = RecordBatch.NO_TIMESTAMP;
+            long batchShallowOffsetOfMaxTimestamp = -1L;
+            // append an entry to the index at batches level (if needed)
+            for (RecordBatch batch : records.batches()) {
+                if (batch.maxTimestamp() > batchMaxTimestamp) {
+                    batchMaxTimestamp = batch.maxTimestamp();
+                    batchShallowOffsetOfMaxTimestamp = batch.lastOffset();
+                }
+
+                // Update the in memory max timestamp and corresponding offset.
+                if (batchMaxTimestamp > maxTimestampSoFar()) {
+                    maxTimestampAndOffsetSoFar = new TimestampOffset(batchMaxTimestamp, batchShallowOffsetOfMaxTimestamp);
+                }
+
+                if (bytesSinceLastIndexEntry > indexIntervalBytes) {
+                    offsetIndex().append(batch.lastOffset(), physicalPosition);
+
+                    // max timestamp may not be monotonic, so we need to check it to avoid the time index append error
+                    if (batchMaxTimestamp >= timeIndex().lastEntry().timestamp)
```

Comment:
This is unnecessary since `TimeIndex` has the same check.

```
            // We only append to the time index when the timestamp is greater than the last inserted timestamp.
            // If all the messages are in message format v0, the timestamp will always be NoTimestamp. In that case, the time
            // index will be empty.
            if (timestamp > lastEntry.timestamp) {

```

Reply:
I think @FrankYang0529 tried to avoid the exception of adding a "smaller" timestamp. see https://github.com/apache/kafka/blob/c28d9a348670f9efc599b481366da0af813aa8ae/storage/src/main/java/org/apache/kafka/storage/internals/log/TimeIndex.java#L199

Reply:
Yes, like @chia7712 mentioned, if there is timestamp index in the file and new timestamp is less than old one, `TimeIndex#maybeAppend` will throw error.